### PR TITLE
69 renewals error msgs

### PR DESCRIPTION
**Pull request by banukutlu**
Issue #69 

### Additional Changes

Refactoring to use bib_summary, added Rails.logger for failed renewals, handled an error when no items selected to renew

### CODE REVIEW

File: app/controllers/renewals_controller.rb [hunk 1]
```diff
@@ -37,7 +38,13 @@ def checkouts_to_renew
...
 
     def renew_params
-      params.require(:renewal_list)
+      if params[:renewal_list].blank?
+        flash[:notice] = 'No items were selected for renewal.'
+
+        redirect_to checkouts_path
+      else
+        params[:renewal_list]
+      end
```
Comment (cdmo) [inline app/controllers/renewals_controller.rb]: I kinda feel like this should be a front-end validate, like unless a checkbox is checked don't allow the button to be clicked (i.e, disable it). What do you think? For the ruby code you can get rid of that `else`.

```suggestion
      end
      params[:renewal_list]

File: app/controllers/renewals_controller.rb [hunk 2]
```diff
@@ -37,7 +38,13 @@ def checkouts_to_renew
     end
 
     def renew_params
-      params.require(:renewal_list)
+      if params[:renewal_list].blank?
+        flash[:notice] = 'No items were selected for renewal.'
+
+        redirect_to checkouts_path
+      else
```
Comment (cdmo) [inline app/controllers/renewals_controller.rb]: I kinda feel like this should be a front-end validate, like unless a checkbox is checked don't allow the button to be clicked (i.e, disable it). What do you think? For the ruby code you can get rid of that else.

```suggestion
      end
      params[:renewal_list]

File: app/controllers/renewals_controller.rb [hunk 3]
```diff
@@ -78,15 +85,18 @@ def bulk_renewal_summary_flash(response, type)
...
-        title_text = "#{renewal[0].title}, #{renewal[0].call_number}"
-        non_renewal_reason = error_prompt(renewal[1])
-        title_text + non_renewal_reason
+        renewal_obj, error_message = renewal
+
+        non_renewal_reason = error_prompt error_message
+        renewal_obj.bib_summary + (non_renewal_reason || '')
       else
-        "#{renewal.title}, #{renewal.call_number}"
+        renewal.bib_summary
```
Comment (cdmo) [inline app/controllers/renewals_controller.rb]: 😎

File: app/models/concerns/bib_record.rb [hunk 4]
```diff
@@ -81,6 +81,10 @@ def item_key
     fields.dig('item', 'key')
   end
 
+  def bib_summary
```
Comment (cdmo) [inline app/models/concerns/bib_record.rb]: ✔️

File: app/controllers/renewals_controller.rb (hunk 3 above)
Comment (cdmo) [reply app/controllers/renewals_controller.rb]: Oh, and again here, you can turn the `else` into an `end` and remove the `end`

File: app/controllers/renewals_controller.rb (hunk 2 above)
Comment (banukutlu) [reply app/controllers/renewals_controller.rb]: yeah, I want to add a front-end validate for this,that's why I brought up the form validations the other day, did not want to add front-end stuff yet before we talk about it. I do not know if it hurts to keep this check as well?

File: app/controllers/renewals_controller.rb (hunk 3 above)
Comment (banukutlu) [reply app/controllers/renewals_controller.rb]: This is for to add the title to the message when there are no errors, the successful renewal case.

Comment (cdmo) [reply app/controllers/renewals_controller.rb]: Yup. So, the way I think of it is the top `if` is catching errors. The default scenario is when renewal.respond_to?(:each) ends up being false. and it's the only other scenario, so you can just remove the else...end and just leave it outside the above if that covers errors

File: app/controllers/renewals_controller.rb (hunk 2 above)
Comment (cdmo) [reply app/controllers/renewals_controller.rb]: Oh sorry, I forgot that conversation. I can see pros and cons of doing it just front-end, just back-end or both. We can chat off GH of course. For now, just backend is in place and working so let's leave it. 👍

File: app/controllers/renewals_controller.rb [hunk 5]
```diff
@@ -73,20 +80,23 @@ def bulk_renewal_flash(response, type)
...
       end
     end
 
     def error_prompt(error_message)
+      return if error_message.empty?
+
       content_tag(:div, "Denied: #{error_message}")
```
Comment (cdmo) [inline app/controllers/renewals_controller.rb]:
```suggestion
      content_tag(:p, "Denied: #{error_message}")
```
Sorry I didn't catch this earlier, but, can we wrap this in [a more semantic](https://developer.mozilla.org/en-US/docs/Learn/Accessibility/HTML) paragraph tag instead? Or does that break things?

File: spec/controllers/checkouts_controller_spec.rb [hunk 6]
```diff
@@ -29,6 +29,12 @@
       allow(mock_patron).to receive(:checkouts).and_return(checkouts)
     end
 
+    it 'sends the right item details to the web service' do
```
Comment (cdmo) [inline spec/controllers/checkouts_controller_spec.rb]: Thanks for adding this 👍

File: spec/controllers/renewals_controller_spec.rb [hunk 7]
```diff
@@ -86,6 +94,16 @@
       end
     end
 
+    context 'when response include errored items with empty error message' do
+      let(:renew_items_response) { { success: [checkouts[0]], error: [[checkouts[2], '']] } }
+
+      it 'error messages include items title only if non renewal reason empty' do
```
Comment (cdmo) [inline spec/controllers/renewals_controller_spec.rb]:
```suggestion
      it 'error messages include items title only' do
```
Seems like `if non renewal reason empty` is redundant?

File: spec/services/symphony_client_spec.rb [hunk 8]
```diff
@@ -100,11 +108,17 @@
                session_token: 'e0b5e1a3e86a399112b9eb893daeacfd')
     end
 
-    it 'returns successful + errored titles for individual renewal requests in symphony' do
-      renew_response = client.renew_items(user, checkouts)
+    it 'returns successful + errored items for individual renewal requests in symphony' do
```
Comment (cdmo) [inline spec/services/symphony_client_spec.rb]:
```suggestion
    it 'returns all responses for individual renewal requests in symphony regardless of success or error' do
```

File: spec/controllers/renewals_controller_spec.rb (hunk 7 above)
Comment (banukutlu) [reply spec/controllers/renewals_controller_spec.rb]: It happens when we get the java errors so I am not adding a detailed msg for the renewal flash, so there should be a div with text starting with `Denied:` in this case, that is what I was trying to test here but not sure not required to do so 🤔

File: app/controllers/renewals_controller.rb (hunk 3 above)
Comment (banukutlu) [reply app/controllers/renewals_controller.rb]: If I understand correctly, you mean sth like this?

``` 
def renewal_prompt(renewal)
      if renewal.respond_to?(:each)
        renewal_obj, error_message = renewal

        non_renewal_reason = error_prompt error_message
        renewal_obj.bib_summary + (non_renewal_reason || '')
      end
    
      renewal.bib_summary
    end
```
For a successful renewal, `renewal` here only has the checkout obj but for failed ones it is an array of checkout and error msg, without the else in the failed renewal case `renewal.bib_summary` won't work 🤔  If I understand correctly or maybe I am confusing myself here :)

File: app/controllers/renewals_controller.rb (hunk 2 above)
Comment (banukutlu) [reply app/controllers/renewals_controller.rb]: yes, let's chat about it and depending on what we decide how we will handle the validations in general, I can refactor this part again or remove etc.

File: spec/services/symphony_client_spec.rb [hunk 9]
```diff
@@ -100,11 +108,17 @@
...
+      renew_response = client.renew_items(user, [checkouts.first, checkouts.second])
 
-      expect(renew_response).to include error: [[checkouts.last, 'Item has holds']],
+      expect(renew_response).to include error: [[checkouts.second, 'Item has holds']],
```
Comment (cdmo) [inline spec/services/symphony_client_spec.rb]:
```suggestion
      expect(renew_response).to be error: [[checkouts.second, 'Item has holds']],
```

The two things being compared have equal (`eq`) values, is this more explicit/specifc?

File: app/controllers/renewals_controller.rb (hunk 5 above)
Comment (banukutlu) [reply app/controllers/renewals_controller.rb]: yes, your suggestion is better 👍

File: spec/controllers/renewals_controller_spec.rb (hunk 7 above)
Comment (cdmo) [reply spec/controllers/renewals_controller_spec.rb]: So, line 97 says `items with empty error message`

I thought that was the same thing?

File: app/controllers/renewals_controller.rb (hunk 3 above)
Comment (banukutlu) [reply app/controllers/renewals_controller.rb]: refactored this per our conversation on slack

File: spec/controllers/renewals_controller_spec.rb (hunk 7 above)
Comment (banukutlu) [reply spec/controllers/renewals_controller_spec.rb]: ahh I see, yes seems redundant

File: app/controllers/renewals_controller.rb [hunk 10]
```diff
@@ -55,38 +62,33 @@ def deny_access
...
-          :ul,
-          safe_join(response[type].map do |renewal|
-            content_tag(:li,
-                        renewal_prompt(renewal),
-                        {},
-                        false)
-          end, '')
-        )
-      )
+      items = response[type].map do |renewal|
```
Comment (cdmo) [inline app/controllers/renewals_controller.rb]: It's even less lines of code! :) Thanks, to me this is so much clearer.

Comment (banukutlu) [reply app/controllers/renewals_controller.rb]: I agree! thanks!

**Pull request merged**